### PR TITLE
[angular] Fix empty selects when the current value is not returned from the server.

### PR DESCRIPTION
**Pull request by mshima**
Fixes https://github.com/jhipster/generator-jhipster/issues/13510

---

Please make sure the below checklist is followed for Pull Requests.

-   [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
-   [ ] Tests are added where necessary
-   [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) was updated if necessary
-   [ ] Documentation is added/updated where necessary
-   [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

**Comment by mshima** [comment]
@kaidohallik what do you think about this one?

**Comment by kaidohallik** [comment]
I have mixed feelings here.

It's definitely good if we fix bug not showing data in the update form.

But current JHipster implementation has limitations which doesn't allow to produce such problematic situation. Maybe correct approach is to improve JHipster relationships related update form logic, like using typeahead with server side search.
Currently every developer must implement it's own logic for selects from paginated entities.

However, if I remember correctly then typeahead has been rejected several times, so probably need to go with this PR way.

I tested this branch. `addUserToCollectionIfMissing` and similar methods in entity side are throwing error if no item is selected. Selected item parameter should be optional.

**Comment by mshima** [comment]
> I have mixed feelings here.
> 
> It's definitely good if we fix bug not showing data in the update form.

That's my opinion too.

> But current JHipster implementation has limitations which doesn't allow to produce such problematic situation. Maybe correct approach is to improve JHipster relationships related update form logic, like using typeahead with server side search.

Not entirely true, just load an external set of data for testing.

> Currently every developer must implement it's own logic for selects from paginated entities.
> 
> However, if I remember correctly then typeahead has been rejected several times, so probably need to go with this PR way.

Just looked for it and seems https://github.com/jhipster/generator-jhipster/pull/6618 was the candidate, but was never finished.
It's not difficult to implement typeahead now using https://ng-bootstrap.github.io/#/components/typeahead/examples.
The missing part is to expose filtering by the otherEntityFieldName.
 
> I tested this branch. `addUserToCollectionIfMissing` and similar methods in entity side are throwing error if no item is selected. Selected item parameter should be optional.

File: generators/client/templates/angular/src/main/webapp/app/core/util/operators.ts.ejs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-%>
+
+export function isPresent<T>(t: T | undefined | null | void): t is T {
+  return t !== undefined && t !== null;
+}
```
Comment (kaidohallik) [inline generators/client/templates/angular/src/main/webapp/app/core/util/operators.ts.ejs:22]: It seems to me that this adds unnecessary complexity. Simple check can be used instead of this added `isPresent`.

File: generators/client/templates/angular/src/main/webapp/app/entities/user/user.model.ts.ejs [hunk 2]
```diff
@@ -20,8 +20,8 @@
   const idType = getTypescriptKeyType(user.primaryKey.type);
 _%>
 export interface IUser {
-  id: <%= idType %>;
-  login: string;
+  id?: <%= idType %>;
+  login?: string;
```
Comment (kaidohallik) [inline generators/client/templates/angular/src/main/webapp/app/entities/user/user.model.ts.ejs:24]: Why this change? When can `id` or `login` can be undefined?

File: generators/client/templates/angular/src/main/webapp/app/entities/user/user.service.ts.ejs [hunk 3]
```diff
@@ -38,4 +39,21 @@ export class UserService {
     const options = createRequestOption(req);
     return this.http.get<IUser[]>(this.resourceUrl, { params: options, observe: 'response' });
   }
+
+  addUserToCollectionIfMissing(userCollection: IUser[], ...usersToCheck: (IUser | null)[]): IUser[] {
+    const users: IUser[] = usersToCheck.filter(isPresent);
```
Comment (kaidohallik) [inline generators/client/templates/angular/src/main/webapp/app/entities/user/user.service.ts.ejs:44]: Can we simplify parameter `usersToCheck` and can we use assignment without `isPresent`?

File: generators/client/templates/angular/src/main/webapp/app/entities/user/user.model.ts.ejs [hunk 4]
```diff
@@ -30,3 +30,7 @@ export class User implements IUser {
     public login: string,
   ) {}
 }
+
+export function getUserIdentifier(user: IUser): <%= idType %> | undefined {
+  return user.id;
+}
```
Comment (kaidohallik) [inline generators/client/templates/angular/src/main/webapp/app/entities/user/user.model.ts.ejs:36]: Why adding this function, why not using just `user.id`?

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/service/entity.service.ts.ejs [hunk 5]
```diff
@@ -99,6 +100,23 @@ export class <%= entityAngularName %>Service {
     }
     <%_ } _%>
 
+    add<%= entityAngularName %>ToCollectionIfMissing(<%= entityInstance %>Collection: I<%= entityAngularName %>[], ...<%= entityInstancePlural %>ToCheck: (I<%= entityAngularName %> | null)[]): I<%= entityAngularName %>[] {
+        const <%= entityInstancePlural %>: I<%= entityAngularName %>[] = <%= entityInstancePlural %>ToCheck.filter(isPresent);
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/service/entity.service.ts.ejs:104]: Can we simplify `ToCheck` parameter and construct assignment without `isPresent`?

**Review by kaidohallik — CHANGES_REQUESTED** [review]

File: generators/client/templates/angular/src/main/webapp/app/entities/user/user.model.ts.ejs (hunk 2 above)
Comment (mshima) [reply generators/client/templates/angular/src/main/webapp/app/entities/user/user.model.ts.ejs:24]: `login` can be undefined using dto:
https://github.com/jhipster/generator-jhipster/blob/ac3f14e815db7e419b4f67541d4c00573aed2bdc/generators/server/templates/src/main/java/package/service/mapper/UserMapper.java.ejs#L142-L152

Not sure about `id`, but I am wondering if it's worth to keep it as non-undefined and keep the quirk only for user entity..

File: generators/client/templates/angular/src/main/webapp/app/entities/user/user.model.ts.ejs (hunk 4 above)
Comment (mshima) [reply generators/client/templates/angular/src/main/webapp/app/entities/user/user.model.ts.ejs:36]: I've added to entities too.
- I have a PR for composite ids that will change every `primaryKey.name` reference.
- Can be reused at `trackEntityByIdentifier` functions.

This should make easier for the upcoming PR, anyway if you prefer to keep references, I'm ok with it.

File: generators/client/templates/angular/src/main/webapp/app/core/util/operators.ts.ejs (hunk 1 above)
Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/core/util/operators.ts.ejs:22]: I tried too now and after playing a bit my conclusion is that your solution is a good solution. I couldn't do better. So marking this and related comments as resolved.

File: generators/client/templates/angular/src/main/webapp/app/entities/user/user.model.ts.ejs (hunk 2 above)
Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/entities/user/user.model.ts.ejs:24]: OK, good point about keeping `User` entity similar to other entities, so this is good to go.

File: generators/client/templates/angular/src/main/webapp/app/entities/user/user.model.ts.ejs (hunk 4 above)
Comment (kaidohallik) [reply generators/client/templates/angular/src/main/webapp/app/entities/user/user.model.ts.ejs:36]: OK, good point about keeping User entity similar to other entities and this way simplify generator, so this is good to go.

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 6]
```diff
@@ -124,6 +129,79 @@ _%>
...
+        describe('Loading relationships', () => {
+            describe('loadRelationshipsOptions', () => {
+    <%_ for (const relationshipsByEntityNeedingOptions of allRelationshipsByEntityNeedingOptions) { _%>
+        <%_ const relationshipsWithCustomUniqueOptions = relationshipsByEntityNeedingOptions.filter(rel => rel.relationshipType === 'one-to-one' && rel.otherEntityName !== 'user'); %>
+        <%_ const relationshipsWithCustomSharedOptions = relationshipsByEntityNeedingOptions.filter(rel => !relationshipsWithCustomUniqueOptions.includes(rel)); %>
+        <%_ if (relationshipsWithCustomSharedOptions.length > 0) { _%>
+            <%_ const otherEntity = relationshipsByEntityNeedingOptions[0].otherEntity _%>
+
+                it('Should call <%= otherEntity.entityAngularName %> query and add missing value', () => {
+                    fakeAsync(() => {
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:143]:
```suggestion
```
If I tested then this code was not executed if using `fakeAsync`. After removing `fakeAsync` test was ran and this test failed.

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 7]
```diff
@@ -124,6 +129,79 @@ _%>
...
+            <%_ for (const relationship of relationshipsWithCustomSharedOptions) { _%>
+                <%_ const reference = relationship.reference _%>
+                            <% if (relationship.collection) { %>...<% } %><%= reference.name %>,
+            <%_ } _%>
+                        ];
+                        const expectedColection: I<%= otherEntity.entityAngularName %>[] = [...additional<%= otherEntity.entityAngularNamePlural %>, ...<%= otherEntity.entityInstance %>Colection];
+                        spyOn(<%= otherEntity.entityInstance %>Service, 'add<%= otherEntity.entityAngularName %>ToCollectionIfMissing').and.returnValues(expectedColection);
+
+                        comp['loadRelationshipsOptions']();
+                        tick();
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:168]:
```suggestion
```

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 8]
```diff
@@ -124,6 +129,79 @@ _%>
...
+
+                        expect(<%= otherEntity.entityInstance %>Service.query).toHaveBeenCalled();
+                        expect(<%= otherEntity.entityInstance %>Service.add<%= otherEntity.entityAngularName %>ToCollectionIfMissing).toHaveBeenCalledWith(<%= otherEntity.entityInstance %>Colection, additional<%= otherEntity.entityAngularNamePlural %>);
+                        expect(comp.<%= otherEntity.entityInstancePlural %>SharedCollection).toEqual(expectedColection);
+                    });
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:173]:
```suggestion
```

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 9]
```diff
@@ -124,6 +129,79 @@ _%>
...
+                        const additional<%= otherEntity.entityAngularNamePlural %> = [
+            <%_ for (const relationship of relationshipsWithCustomSharedOptions) { _%>
+                <%_ const reference = relationship.reference _%>
+                            <% if (relationship.collection) { %>...<% } %><%= reference.name %>,
+            <%_ } _%>
+                        ];
+                        const expectedColection: I<%= otherEntity.entityAngularName %>[] = [...additional<%= otherEntity.entityAngularNamePlural %>, ...<%= otherEntity.entityInstance %>Colection];
+                        spyOn(<%= otherEntity.entityInstance %>Service, 'add<%= otherEntity.entityAngularName %>ToCollectionIfMissing').and.returnValues(expectedColection);
+
+                        comp['loadRelationshipsOptions']();
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:167]:
```suggestion
                        comp.ngOnInit();
```
It's a good practice to test public interface. Can you give input to component through `ActivatedRoute` and run public `ngOnInit()` instead?

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 10]
```diff
@@ -124,6 +129,79 @@ _%>
...
+            <%_ for (const relationship of relationshipsWithCustomSharedOptions) { _%>
+                <%_ const reference = relationship.reference _%>
+                <%_ if (relationship.collection) { _%>
+                        const <%= reference.name %> : I<%= otherEntity.entityAngularName %>[] = [<%- generateTestEntityPrimaryKey(otherEntity.primaryKey) %>];
+                <%_ } else { _%>
+                        const <%= reference.name %> : I<%= otherEntity.entityAngularName %> = <%- generateTestEntityPrimaryKey(otherEntity.primaryKey) %>;
+                <%_ } _%>
+                        <%= entityInstance %>.<%= reference.name %> = <%= reference.name %>;
+            <%_ } _%>
+                        comp.updateForm(<%= entityInstance %>);
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:154]: Can you give input through `ActivatedRoute` and then this functionality is called through `ngOnInit()`.
Method `updateForm` should be protected in component. If you agree to make this change as well in this PR then it would be great. If not then this change can be done in a separate PR.

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 11]
```diff
@@ -124,6 +129,79 @@ _%>
...
+                        expect(comp.<%= otherEntity.entityInstancePlural %>SharedCollection).toEqual(expectedColection);
+                    });
+                });
+        <%_ } _%>
+        <%_ for (const relationship of relationshipsWithCustomUniqueOptions) { %>
+            <%_ const otherEntity = relationship.otherEntity _%>
+            <%_ const reference = relationship.reference _%>
+
+                it('Should call <%= reference.name %> query and add missing value', () => {
+                    fakeAsync(() => {
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:181]:
```suggestion
```
Test is not executed with `fakeAsync`. After removing test is failing.

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 12]
```diff
@@ -124,6 +129,79 @@ _%>
...
+                        const <%= entityInstance %> : I<%= entityAngularName %> = <%- generateTestEntityPrimaryKey(primaryKey, 0) %>;
+                        const <%= reference.name %> : I<%= otherEntity.entityAngularName %> = <%- generateTestEntityPrimaryKey(otherEntity.primaryKey) %>;
+                        <%= entityInstance %>.<%= reference.name %> = <%= reference.name %>;
+                        comp.updateForm(<%= entityInstance %>);
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:185]: Give input through `ActivatedRoute`.

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 13]
```diff
@@ -124,6 +129,79 @@ _%>
...
+
+                        const <%= reference.name %>Colection: I<%= otherEntity.entityAngularName %>[] = [<%- generateTestEntityPrimaryKey(otherEntity.primaryKey) %>];
+                        spyOn(<%= otherEntity.entityInstance %>Service, 'query').and.returnValue(of(new HttpResponse({ body: <%= reference.name %>Colection })));
+                        const expectedColection: I<%= otherEntity.entityAngularName %>[] = [<%= reference.name %>, ...<%= reference.name %>Colection];
+                        spyOn(<%= otherEntity.entityInstance %>Service, 'add<%= otherEntity.entityAngularName %>ToCollectionIfMissing').and.returnValue(expectedColection);
+
+                        comp['loadRelationshipsOptions']();
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:192]:
```suggestion
                        comp.ngOnInit();
```
Give input through `ActivatedRoute`.

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 14]
```diff
@@ -124,6 +129,79 @@ _%>
...
+                        tick();
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:193]:
```suggestion
```

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 15]
```diff
@@ -124,6 +129,79 @@ _%>
...
+
+                        expect(<%= otherEntity.entityInstance %>Service.query).toHaveBeenCalled();
+                        expect(<%= otherEntity.entityInstance %>Service.add<%= otherEntity.entityAngularName %>ToCollectionIfMissing).toHaveBeenCalledWith(<%= reference.name %>Colection, <%= reference.name %>);
+                        expect(comp.<%= relationship.relationshipFieldNamePlural %>Collection).toEqual(expectedColection);
+                    });
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:198]:
```suggestion
```

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 16]
```diff
@@ -124,6 +129,79 @@ _%>
...
+                        const <%= reference.name %> : I<%= otherEntity.entityAngularName %>[] = [<%- generateTestEntityPrimaryKey(otherEntity.primaryKey) %>];
+                <%_ } else { _%>
+                        const <%= reference.name %> : I<%= otherEntity.entityAngularName %> = <%- generateTestEntityPrimaryKey(otherEntity.primaryKey) %>;
+                <%_ } _%>
+                        <%= entityInstance %>.<%= reference.name %> = <%= reference.name %>;
+            <%_ } _%>
+                        comp.updateForm(<%= entityInstance %>);
+
+                        const <%= otherEntity.entityInstance %>Colection: I<%= otherEntity.entityAngularName %>[] = [<%- generateTestEntityPrimaryKey(otherEntity.primaryKey) %>];
+                        spyOn(<%= otherEntity.entityInstance %>Service, 'query').and.returnValue(of(new HttpResponse({ body: <%= otherEntity.entityInstance %>Colection })));
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:157]: Typo, use Collection instead of Colection.

File: generators/client/templates/angular/src/main/webapp/app/core/util/operators.ts.ejs (hunk 1 above)
Comment (kaidohallik) [inline generators/client/templates/angular/src/main/webapp/app/core/util/operators.ts.ejs:22]: Can you add test to this method?

File: generators/client/templates/angular/src/main/webapp/app/entities/user/user.service.spec.ts.ejs [hunk 17]
```diff
@@ -81,6 +78,42 @@ describe('Service Tests', () => {
...
+            {
+              ...user,
+            },
+            <%- generateTestEntityPrimaryKey(user.primaryKey, 1) %>,
+          ];
+          expectedResult = service.addUserToCollectionIfMissing(userCollection, user);
+          expect(expectedResult).toHaveLength(2);
+        });
+
+        it("should add a User to an array that doesn't contains it", () => {
```
Comment (kaidohallik) [inline generators/client/templates/angular/src/main/webapp/app/entities/user/user.service.spec.ts.ejs:102]:
```suggestion
        it("should add a User to an array that doesn't contain it", () => {
```

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/service/entity.service.spec.ts.ejs [hunk 18]
```diff
@@ -235,6 +235,55 @@ describe('Service Tests', () => {
...
+                it('should not add a <%= entityAngularName %> to an array that contains it', () => {
+                    const <%= entityInstance %>: I<%= entityAngularName %> = <%- generateTestEntityPrimaryKey(primaryKey, 0) %>;
+                    const <%= entityInstance %>Collection: I<%= entityAngularName %>[] = [{
+                        ...<%= entityInstance %>,
+                    }, <%- generateTestEntityPrimaryKey(primaryKey, 1) %>];
+                    expectedResult = service.add<%= entityAngularName %>ToCollectionIfMissing(<%= entityInstance %>Collection, <%= entityInstance %>);
+                    expect(expectedResult).toHaveLength(2);
+                });
+
+                it("should add a <%= entityAngularName %> to an array that doesn't contains it", () => {
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/service/entity.service.spec.ts.ejs:256]:
```suggestion
                it("should add a <%= entityAngularName %> to an array that doesn't contain it", () => {
```

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/service/entity.service.spec.ts.ejs [hunk 19]
```diff
@@ -235,6 +235,55 @@ describe('Service Tests', () => {
...
+                    const <%= entityInstance %>2: I<%= entityAngularName %> = <%- generateTestEntityPrimaryKey(primaryKey, 1) %>;
+                    expectedResult = service.add<%= entityAngularName %>ToCollectionIfMissing([], <%= entityInstance %>, <%= entityInstance %>2);
+                    expect(expectedResult).toHaveLength(2);
+                    expect(expectedResult).toContain(<%= entityInstance %>);
+                    expect(expectedResult).toContain(<%= entityInstance %>2);
+                });
+
+                it("should accept null values", () => {
+                    const <%= entityInstance %>: I<%= entityAngularName %> = <%- generateTestEntityPrimaryKey(primaryKey, 0) %>;
+                    expectedResult = service.add<%= entityAngularName %>ToCollectionIfMissing([], null, <%= entityInstance %>);
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/service/entity.service.spec.ts.ejs:282]:
```suggestion
                it("should accept null and undefined values", () => {
                    const <%= entityInstance %>: I<%= entityAngularName %> = <%- generateTestEntityPrimaryKey(primaryKey, 0) %>;
                    expectedResult = service.add<%= entityAngularName %>ToCollectionIfMissing([], null, <%= entityInstance %>, undefined);
```

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/service/entity.service.spec.ts.ejs [hunk 20]
```diff
@@ -235,6 +235,55 @@ describe('Service Tests', () => {
...
+                });
+
+                it("should add only unique <%= entityAngularName %> to an array", () => {
+                    const <%= entityInstance %>Array: I<%= entityAngularName %>[] = [<%- generateTestEntityPrimaryKey(primaryKey, 0) %>, <%- generateTestEntityPrimaryKey(primaryKey, 1) %>, <%- generateTestEntityPrimaryKey(primaryKey) %>];
+                    const <%= entityInstance %>Collection: I<%= entityAngularName %>[] = [<%- generateTestEntityPrimaryKey(primaryKey, 1) %>];
+                    expectedResult = service.add<%= entityAngularName %>ToCollectionIfMissing(<%= entityInstance %>Collection, ...<%= entityInstance %>Array);
+                    expect(expectedResult).toHaveLength(3);
+                });
+
+                it("should accept varargs", () => {
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/service/entity.service.spec.ts.ejs:271]: It's a good test. Can you add varargs test also for `user.service`?

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/service/entity.service.spec.ts.ejs [hunk 21]
```diff
@@ -235,6 +235,55 @@ describe('Service Tests', () => {
...
+                    const <%= entityInstance %>: I<%= entityAngularName %> = <%- generateTestEntityPrimaryKey(primaryKey, 0) %>;
+                    const <%= entityInstance %>2: I<%= entityAngularName %> = <%- generateTestEntityPrimaryKey(primaryKey, 1) %>;
+                    expectedResult = service.add<%= entityAngularName %>ToCollectionIfMissing([], <%= entityInstance %>, <%= entityInstance %>2);
+                    expect(expectedResult).toHaveLength(2);
+                    expect(expectedResult).toContain(<%= entityInstance %>);
+                    expect(expectedResult).toContain(<%= entityInstance %>2);
+                });
+
+                it("should accept null values", () => {
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/service/entity.service.spec.ts.ejs:280]: It's a good test. Can you add null + undefined test also for `user.service`?

**Review by kaidohallik — CHANGES_REQUESTED** [review]
Now console error is gone and code is working. Good work @mshima!
I have some comments about polishing and tests.

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/service/entity.service.spec.ts.ejs (hunk 20 above)
Comment (mshima) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/service/entity.service.spec.ts.ejs:271]: Done

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/service/entity.service.spec.ts.ejs (hunk 21 above)
Comment (mshima) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/service/entity.service.spec.ts.ejs]: Done

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs (hunk 9 above)
Comment (mshima) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:166]: Not an expert on angular tests.
So can take some time.

Comment (kaidohallik) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:166]: Sample of providing `ActivatedRoute` is for example in https://github.com/jhipster/generator-jhipster/blob/main/generators/entity-client/templates/angular/src/main/webapp/app/entities/detail/entity-management-detail.component.spec.ts.ejs#L44-L49

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 22]
```diff
@@ -26,64 +30,183 @@ import { HttpResponse } from '@angular/common/http';
...
 
 import { <%= entityAngularName %>UpdateComponent } from './<%= entityFileName %>-update.component';
 
 describe('Component Tests', () => {
     describe('<%= entityAngularName %> Management Update Component', () => {
+        const activatedRouteData = new Subject();
+        const activatedRouteStub: any = {
+            data: activatedRouteData.asObservable(),
+        };
+
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:57]:
```suggestion
        let activatedRouteData: Observable<I<%= entityAngularName %>>;
```
I think it's more natural to construct desired `ActivatedRoute` and then call `comp.ngOnInit();` as this is happening in real workflow.

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 23]
```diff
@@ -26,64 +30,183 @@ import { HttpResponse } from '@angular/common/http';
...
         beforeEach(() => {
             TestBed.configureTestingModule({
                 imports: [HttpClientTestingModule],
                 declarations: [<%= entityAngularName %>UpdateComponent],
-                providers: [FormBuilder, ActivatedRoute]
+                providers: [
+                    FormBuilder,
+                    {
+                        provide: ActivatedRoute,
+                        useValue: activatedRouteStub,
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:74]:
```suggestion
                        useValue: { data: activatedRouteData },
```

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 24]
```diff
@@ -26,64 +30,183 @@ import { HttpResponse } from '@angular/common/http';
...
+                        const <%= reference.name %> : I<%= otherEntity.entityAngularName %> = <%- generateTestEntityPrimaryKey(otherEntity.primaryKey) %>;
+                        <%= entityInstance %>.<%= reference.name %> = <%= reference.name %>;
+
+                        const <%= reference.name %>Collection: I<%= otherEntity.entityAngularName %>[] = [<%- generateTestEntityPrimaryKey(otherEntity.primaryKey) %>];
+                        spyOn(<%= otherEntity.entityInstance %>Service, 'query').and.returnValue(of(new HttpResponse({ body: <%= reference.name %>Collection })));
+                        const expectedCollection: I<%= otherEntity.entityAngularName %>[] = [<%= reference.name %>, ...<%= reference.name %>Collection];
+                        spyOn(<%= otherEntity.entityInstance %>Service, 'add<%= otherEntity.entityAngularName %>ToCollectionIfMissing').and.returnValue(expectedCollection);
+
+                        // Trigger an ActivatedRoute change
+                        activatedRouteData.next({ <%= entityInstance %> });;
```
Comment (kaidohallik) [inline generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:161]:
```suggestion
                        activatedRouteData = of({ <%= entityInstance %> });
                        comp.ngOnInit();
```

**Review by kaidohallik — CHANGES_REQUESTED** [review]
I made some suggestions to use `ngOnInit();` instead of `updateForm`.

File: generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs [hunk 25]
```diff
@@ -26,64 +30,183 @@ import { HttpResponse } from '@angular/common/http';
...
+                const <%= entityInstance %>: I<%= entityAngularName %> = <%- generateTestEntityPrimaryKey(primaryKey, 1) %>;
+<%_ for (const relationshipsByEntityNeedingOptions of allRelationshipsByEntityNeedingOptions) { _%>
+    <%_ for (const relationship of relationshipsByEntityNeedingOptions) { _%>
+        <%_ const otherEntity = relationship.otherEntity _%>
+                const <%= relationship.reference.name %>: I<%= otherEntity.entityAngularName %> = <%- generateTestEntityPrimaryKey(otherEntity.primaryKey) %>;
+                <%= entityInstance %>.<%= relationship.reference.name %> = <%= relationship.collection ? `[${relationship.reference.name}]` : relationship.reference.name %>;
+    <%_ } _%>
+<%_ } _%>
+
+                comp.updateForm(<%= entityInstance %>);
```
[the remark this answers is not in the document]
Comment (mshima) [reply generators/entity-client/templates/angular/src/main/webapp/app/entities/update/entity-management-update.component.spec.ts.ejs:184]:
```suggestion
                activatedRouteData = of({ <%= entityInstance %> });
                comp.updateForm(<%= entityInstance %>);
```

**Comment by mshima** [comment]
@kaidohallik I've tried your approach, but `useValue` is created at beforeEach and `activatedRouteData = of({ <%= entityInstance %> });` is only applied at next test (beforeEach).

The Subject approach creates a dynamic Observable, so when `activatedRouteData.next({ <%= entityInstance %> });` is called, the route subscription is triggered.

Others alternatives are:
- Use only one `<%= entityInstance %>` for every test.
- Create `before` steps to set the Observable instance.

Current Subject approach seems better to me.

**Review by kaidohallik — APPROVED** [review]
This PR is good to go.
Thanks @mshima
We can do follow up PR if we want to improve entity update test.

**Comment by mshima** [comment]
@kaidohallik thanks
